crates/sage/src/idpicker.rs

```diff
@@ -0,0 +1,563 @@
+                } else if protein_map_pg2.contains_key(each_protein) {
+                    // tier-2 proteins groups
+                    protein_map_pg2.get(each_protein).unwrap().join("/")
+                } else {
+                    // If the protein is not found in either tiers, return an error
+                    // panic!("{}",format!("Protein {} not found in protein group!!", each_protein));
+                    each_protein.to_string()
+                }
+            })
+            .collect();
+
+        feat.proteingroups = Some(proteingroups.iter().sorted().join(";"));
+        feat.num_proteingroups = proteingroups.len() as i32;
+    });
+}
+
+fn get_peptide_protein_map(
+    db: &&IndexedDatabase,
+    features: &mut [Feature],
+    gp1_filter: fn(i32, f32) -> bool,
+) -> Vec<(Arc<String>, Arc<String>)> {
+    features
+        .iter()
+        .filter(|feature: &&Feature| gp1_filter(feature.label, feature.peptide_q))
+        .flat_map(|feature| {
+            let idx = feature.peptide_idx;
+            let proteins = &db[feature.peptide_idx].proteins(&db.decoy_tag, db.generate_decoys);
+            let peptide_proteins = proteins
+                .split(";")
+                .map(|s| (Arc::new(idx.0.to_string()), Arc::new(s.to_string())))
+                .collect_vec();
+            peptide_proteins
+        })
+        .collect_vec()
+}
+
+fn get_proteingroups(
+    pep_proteins: Vec<(Arc<String>, Arc<String>)>,
+) -> FxHashMap<String, Vec<String>> {
+    info!("- number of records: {}", pep_proteins.len());
+    let start = Instant::now();
+
+    let start_meth = Instant::now();
+    let group_by_peptides = get_group_by_peptides(pep_proteins);
+    let end_time = Instant::now() - start_meth;
+    info!("-  get_group_by_peptides: {:8} ms", end_time.as_millis());
+
+    let start_meth = Instant::now();
+    let grouped_peptides_and_grouped_proteins =
+        get_grouped_peptides_and_proteins(group_by_peptides);
+    let end_time = Instant::now() - start_meth;
+    info!("-  get_grouped_peptides_and_proteins: {:8} ms", end_time.as_millis());
+
+    let start_meth = Instant::now();
+    let final_cluster = separate_into_clusters(grouped_peptides_and_grouped_proteins);
+    let end_time = Instant::now() - start_meth;
+    info!("-  separate_into_clusters: {:8} ms", end_time.as_millis());
+
+    let final_proteins: Vec<(String, Vec<String>)> = final_cluster
+        .into_par_iter()
+        .flat_map(|cm| reduce_cluster(cm.1))
+        .collect();
+    let protein_map: FxHashMap<_, _> = final_proteins.into_par_iter().collect();
+
+    let protein_map_process_time = Instant::now() - start;
+    info!("- proteingroup: {:8} ms", protein_map_process_time.as_millis());
+    protein_map
+}
+
+pub fn get_group_by_peptides(
+    data: Vec<(Arc<String>, Arc<String>)>,
+) -> FxHashSet<(Vec<Arc<String>>, Arc<String>)> {
+    // Phase 1: Group proteins by peptide
+    let group_by_peptides: FxHashMap<Arc<String>, Vec<Arc<String>>> = {
```

Comment:
Is there a reason we can't use `PeptideIx` instead of `Arc<String>` to represent peptides?